pandas/_libs/window_indexer.pyx

```diff
@@ -0,0 +1,167 @@
+# cython: boundscheck=False, wraparound=False, cdivision=True
+
+import numpy as np
+from numpy cimport ndarray, int64_t
+
+# ----------------------------------------------------------------------
+# The indexer objects for rolling
+# These define start/end indexers to compute offsets
+
+
+class MockFixedWindowIndexer:
+    """
+
+    We are just checking parameters of the indexer,
+    and returning a consistent API with fixed/variable
+    indexers.
+
+    Parameters
+    ----------
+    values: ndarray
+        values data array
+    win: int64_t
+        window size
+    index: object
+        index of the values
+    closed: string
+        closed behavior
+    """
+    def __init__(self, ndarray values, int64_t win, object closed, object index=None):
+
+        self.start = np.empty(0, dtype='int64')
+        self.end = np.empty(0, dtype='int64')
+
+    def get_window_bounds(self):
+        return self.start, self.end
+
+
+class FixedWindowIndexer:
+    """
+    create a fixed length window indexer object
+    that has start & end, that point to offsets in
+    the index object; these are defined based on the win
+    arguments
+
+    Parameters
+    ----------
+    values: ndarray
+        values data array
+    win: int64_t
+        window size
+    index: object
+        index of the values
+    closed: string
+        closed behavior
+    """
+    def __init__(self, ndarray values, int64_t win, object closed, object index=None):
+        cdef:
+            ndarray start_s, start_e, end_s, end_e
+            int64_t N = len(values)
+
+        start_s = np.zeros(win, dtype='int64')
+        start_e = np.arange(win, N, dtype='int64') - win + 1
+        self.start = np.concatenate([start_s, start_e])
+
+        end_s = np.arange(win, dtype='int64') + 1
+        end_e = start_e + win
+        self.end = np.concatenate([end_s, end_e])
+
+    def get_window_bounds(self):
```

Comment:
can you type

Reply:
Python typing? These are defined in cython files still.

Reply:
hmm interesting, though I *think* still can type. Anyhow for a try in a followup.

Reply:
on non-cdef/cpdef functions you can, no problem.  i dont think it optimizes anything though

Reply:
its optimizing readability :->